File: Leetcode/Python/_198.py
class Solution:
    # Constant space
    def rob(self, nums: List[int]) -> int:
        n = len(nums)
        if n == 1:
            return nums[0]
        first, second = nums[0], max(nums[0], nums[1])
        for i in range(2, n):
            first, second = second, max(second, first + nums[i])
        return second
      
class Solution:
    # dp with linear space 
    def rob(self, nums: List[int]) -> int:
        n = len(nums)
        dp = [0] * (n + 2)
        for i in range(2, n + 2):
            dp[i] = max(dp[i - 1], dp[i - 2] + nums[i - 2])
        return dp[-1]

File: Leetcode/cpp/198.cpp
#include <iostream>
#include <vector>
#include <cassert>

class Solution {
public:
  int rob(std::vector<int>& nums) {
    int n = static_cast<int>(nums.size());
    std::vector<int> dp(n + 2, 0);
    for (int i = 0; i < n; ++i) {
      dp[i + 2] = std::max(dp[i] + nums[i], dp[i + 1]);
    }
    return dp[n + 1];
  }
};

int main() {
  Solution sol;
  std::vector input = {1, 2, 3, 1};
  int expected_output = 4;
  
  int output = sol.rob(input);
  assert (output == expected_output && "Test #1 failed!");
  std::cout << "Tests are passed!" << std::endl;
  
  return 0;
}
